documentation/API-GUIDELINES.md

```diff
@@ -15,16 +15,28 @@ The following paragraphs contain additional recommendations.
 
 ## Construction and Destruction of Drivers
 
-- Drivers take peripherals and pins via the `PeripheralRef` pattern - they don't consume peripherals/pins.
+- Drivers should take peripherals via the `PeripheralRef` pattern - they don't consume peripherals directly.
+- If a driver requires pins, those pins should be configured using `fn with_pin_name(self, pin: impl Peripheral<P = InputConnection> + 'd) -> Self)` or `fn with_pin_name(self, pin: impl Peripheral<P = OutputConnection> + 'd) -> Self)`
+- If a driver supports multiple peripheral instances:
+  - The peripheral instance type should be positioned as the last type parameter of the driver type.
+  - The peripheral instance type should default to a type that supports any of the peripheral instances.
+  - The author is encouraged to use `crate::any_peripheral` to define the "any" peripheral instance type.
+  - The driver should implement a `new` constructor that automatically converts the peripheral instance into the any type, and a `new_typed` that preserves the peripheral type.
+- If a driver implements both blocking and async operations, or only implements blocking operations, but may support asynchronous ones in the future, the driver's type signature should include a `crate::Mode` type parameter.
+- By default, constructors should configure the driver for blocking mode. The driver should implement `into_async` (and a matching `into_blocking`) function that reconfigures the driver.
+  - `into_async` should configure the driver and/or the associated DMA channels. This most often means enabling an interrupt handler.
```

Comment:
I still have mixed feelings about enforcing this here without first hashing out how this should work properly

Reply:
Yeah but I was asked to do this 🤷‍♂️